### PR TITLE
Add recipe for meson-python

### PR DESCRIPTION
**Pull request by awvwgk**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by awvwgk** [comment]
@conda-forge/help-python This recipe is ready for review. While I would prefer `noarch: python`, it seems more important to support all Python versions to not limit other Python packages planning to use this one as build system (like SciPy).

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks for working on this @awvwgk!

The patch here looks wrong to me. If there's something to fix, let's do it upstream. But the reason for changing `ninja` requirements here is not correct either way: when one uses build isolation in a Python venv, a `ninja` installed by conda-forge cannot be used.

This package should be `noarch`. The `pip check` failure may be due to incorrect metadata in `pyproject.toml`; if needed a patch can be carried for that until it's fixed in the next `meson-python` release.

File: recipes/meson-python/meta.yaml
Comment (rgommers) [inline recipes/meson-python/meta.yaml:28]: `patchelf` is also needed, see `mesonpy/_elf.py`.

Comment (rgommers) [reply recipes/meson-python/meta.yaml:28]: On Linux only

Comment (awvwgk) [reply recipes/meson-python/meta.yaml:28]: Since we are building for noarch now tests will only run on Linux.

Comment (rgommers) [reply recipes/meson-python/meta.yaml:28]: There's actually a good argument for leaving it out, but I'm not sure if it's currently optional. It will depend on whether there's a standalone non-Python library being vendored.

Actually, on second though - maybe fine to leave it out until someone needs it. WDYT?

Comment (awvwgk) [reply recipes/meson-python/meta.yaml:28]: We cannot depend on platform specific libraries with a noarch package, if platform specific dependencies are required, this package cannot be noarch.

Comment (rgommers) [reply recipes/meson-python/meta.yaml]: Ah, then let's leave it out - SciPy doesn't need it, and I don't think most other packages need it either.

Comment (carterbox) [reply recipes/meson-python/meta.yaml]: > We cannot depend on platform specific libraries with a noarch package, if platform specific dependencies are required, this package cannot be noarch.

That's [technically not true](https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python). If you have a dependency that is available only available for some platforms, we still recommend noarch it's just that the package will be uninstallable for platforms which do not have the dependency.

There are two ways to remedy this:

1. If the dependency is required, accept that the package is uninstallable on some platforms because it doesn't work on those platforms but the package itself is pure python.
2. If the dependency is optional, you can remove the package from the run to run_constrained and users need to manually install if they want. OR ask upstream to add an empty metapackage for the platforms that are missing the package.

Comment (awvwgk) [reply recipes/meson-python/meta.yaml]: Well, that's not what I meant, but point taken ;). I'll switch to noarch and drop patchelf from the run requirements.

**Comment by awvwgk** [comment]
Seems like this is ready for review again, now as `noarch; python`.

File: recipes/meson-python/meta.yaml
Comment (rgommers) [inline recipes/meson-python/meta.yaml:64]: Can you add me too @awvwgk? SciPy builds will depend on `meson-python`, so I'm interested in helping maintain this package.

File: recipes/meson-python/ninja-overdep.patch
Comment (rgommers) [inline recipes/meson-python/ninja-overdep.patch:9]: This patch is kinda unusual, so let me summarize my understanding:
- `PKG-INFO` and `pyproject.toml` are not used at all by `conda` tooling, so no harm done there
- This is added to make `pip check` in CI happy

Maybe worth adding a comment if we're keeping this patch permanently?

**Review by rgommers — APPROVED** [review]
LGTM overall now, thanks @awvwgk

File: recipes/meson-python/meta.yaml
Comment (awvwgk) [reply recipes/meson-python/meta.yaml:64]: Sure

File: recipes/meson-python/ninja-overdep.patch
Comment (awvwgk) [reply recipes/meson-python/ninja-overdep.patch:9]: I'm not an expert in Python package metadata, but at least the `project.dependencies` list in `pyproject.toml` creates incorrect information on the requirements of this package, which will be present in the installed package and are therefore flagged by `pip`.

The `Requires-Dist:` tag and `build-system.requires` entries are probably less critical here, but are patched for consistency to make the intent.

Patching dependency metadata in projects from PyPI is not unusual IMO, at least for the projects I packaged so far I found it to be incorrect or incomplete in many cases.

Comment (rgommers) [reply recipes/meson-python/ninja-overdep.patch:9]: I can't say I understand this. `ninja` _is_ a required dependency. Beyond avoiding `pip check`, I'm not sure what this patch does or incorrect information means. Maybe one of the other reviewers knows?

Comment (awvwgk) [reply recipes/meson-python/ninja-overdep.patch:9]: I guess the question boils down to whether `pip` is supposed to manage the dependency on a C++ program or not.

Comment (rgommers) [reply recipes/meson-python/ninja-overdep.patch:9]: Ah, I guess we just came to the same conclusion of that being the question:)  https://github.com/FFY00/meson-python/issues/60#issuecomment-1133957875

Comment (BastianZim) [reply recipes/meson-python/ninja-overdep.patch:9]: Just as an fyi mostly: Pip check is not necessary. It’s nice to have but fails in many situations where a dependency doesn‘t properly declare itself (like here). 
I‘d be happy to merge the PR without it if that's easier for you.

Comment (awvwgk) [reply recipes/meson-python/ninja-overdep.patch:9]: I would prefer to keep this patch to have tools like `build`, `pip` or `wheel` which work with the package metadata not flag the unrecognized dependency from the conda package. Of course it is possible to manually silence those warnings and checks, but this moves the burden to the user.

File: recipes/meson-python/meta.yaml
Comment (rgommers) [inline recipes/meson-python/meta.yaml:23]: I think `wheel` should be here; must have overlooked that before. It gets pulled in anyway via `python` I guess, but it's a direct dependency so good to capture.

File: recipes/meson-python/ninja-overdep.patch
Comment (rgommers) [reply recipes/meson-python/ninja-overdep.patch:9]: 0.5.0 should be released later today with some changes to dependencies, may be worth waiting for that and updating this tomorrow.

Comment (awvwgk) [reply recipes/meson-python/ninja-overdep.patch:9]: Updated to 0.5.0

**Review by rgommers — APPROVED** [review]
LGTM, thanks!

**Comment by rgommers** [comment]
@conda-forge/staged-recipes can this be merged?

**Pull request merged**
